inc/composer/class-plugin.php

```diff
@@ -51,8 +56,126 @@ public static function getSubscribedEvents() {
 	 * in addition to the $this->activate() method.
 	 */
 	public function init() {
-		$installer = new Installer( $this->io, $this->composer );
-		$this->composer->getInstallationManager()->addInstaller( $installer );
+		$this->installer = new Installer( $this->io, $this->composer );
+		$this->composer->getInstallationManager()->addInstaller( $this->installer );
+	}
+
+	/**
+	 * Update install overrides once dependencies have been resolved. (Composer v2)
+	 *
+	 * @param InstallerEvent $event
+	 * @return void
+	 */
+	public function pre_operations_exec( InstallerEvent $event )  {
+		$transaction = $event->getTransaction();
+		$operations = $transaction->getOperations();
+		if ( empty( $operations ) ) {
+			return;
+		}
+
+		// Work out which packages we already have.
+		$repo = $event->getComposer()->getLocker()->getLockedRepository( $event->isDevMode() );
+		$packages = [];
+		foreach ( $repo->getPackages() as $package ) {
+			$packages[ $package->getName() ] = $package;
+		}
+
+		// Then, resolve the operations we're about to apply.
+		// (In Composer v2, this is when running the initial install step.)
+		foreach ( $operations as $operation ) {
+			switch ( $operation::TYPE ) {
+				case 'install':
+				case 'markAliasInstalled':
+					$package = $operation->getPackage();
+					$packages[ $package->getName() ] = $package;
+					break;
+
+				case 'update':
+					$package = $operation->getTargetPackage();
+					$packages[ $package->getName() ] = $package;
+					break;
+
+				case 'uninstall':
+				case 'markAliasUninstalled':
+					$package = $operation->getPackage();
+					unset( $packages[ $package->getName() ] );
+					break;
+
+				default:
+					break;
+			}
+		}
+
+		$overrides = $this->getAllInstallOverrides( $packages );
+		$this->installer->setInstallOverrides( $overrides );
+	}
+
+	/**
+	 * Update install overrides once dependencies have been resolved. (Composer v1)
+	 *
+	 * @param InstallerEvent $event
+	 * @return void
+	 */
+	public function post_dependencies_solving( InstallerEvent $event )  {
+		$operations = $event->getOperations();
+		if ( empty( $operations ) ) {
+			return;
+		}
+
+		// First, work out which packages we already have.
+		$repo = $event->getInstalledRepo();
```

Comment:
Given this is the only difference between the methods could the hook be derived from the `InstallerEvent` object instead?

Reply:
Possibly; I'd rather have a common method used by both handlers here instead, as I'm wary of heuristically detecting too much. That said, I left that part until I fix the behaviour first